xml/clp/src/main/scala/org/broadinstitute/monster/extractors/xml/ExtractorClp.scala

```diff
@@ -0,0 +1,103 @@
+import cats.effect.{ExitCode, IO, Resource}
+import cats.implicits._
+import com.monovore.decline._
+import com.monovore.decline.effect._
+
+import scala.concurrent.ExecutionContext
+
+/**
+  * Command-line program which can use our extractor functionality to convert
+  * XML to chunked JSON on local disk.
+  */
+object ExtractorClp
+    extends CommandIOApp(
+      name = "xml-extract",
+      header = "Mechanically extract XML data into JSON-list",
+      version = BuildInfo.version
+    ) {
+
+  implicit val fileArg: Argument[File] = new Argument[File] {
+    override def read(string: String): ValidatedNel[String, File] =
+      try {
+        Validated.validNel(File(string))
+      } catch {
+        case e: Exception =>
+          Validated.invalidNel(s"Failed to parse '$string' as a path: ${e.getMessage}")
+      }
+
+    override def defaultMetavar: String = "path"
+  }
+
+  override def main: Opts[IO[ExitCode]] = {
+    val inputOpt = Opts
+      .option[File](
+        "input",
+        "Path to the local XML file to extract.",
+        short = "i"
+      )
+      .validate("Input does not exist")(_.isReadable)
+
+    val outputOpt = Opts
+      .option[File](
+        "output",
+        "Path to the local directory where extracted JSON should be written.",
+        short = "o"
+      )
+      .validate("Output is not a directory") { out =>
+        out.notExists || out.isDirectory
+      }
+
+    val gzipOpt = Opts.flag("gunzip-input", "Gunzip input before extraction").orFalse
+
+    val tagOpt = Opts.option[String](
+      "xml-tag",
+      "XML tag to extract as top-level objects in the JSON-list output.",
+      short = "t"
+    )
+
+    val countOpt = Opts.option[Int](
+      "objects-per-part",
+      "Number of JSON objects to write to each partfile in the output.",
+      short = "n"
+    )
+
+    (inputOpt, outputOpt, gzipOpt, tagOpt, countOpt).mapN {
+      case (in, out, gunzip, tag, count) =>
+        val blockingEc = {
+          val allocate = IO.delay(Executors.newCachedThreadPool())
+          val free = (es: ExecutorService) => IO.delay(es.shutdown())
+          Resource.make(allocate)(free).map(ExecutionContext.fromExecutor)
+        }
+
+        blockingEc.use { ec =>
+          val extractor = {
+            val readPath = (inFile: File) => {
+              val base = fs2.io.file.readAll[IO](inFile.path, ec, 8196)
+              if (gunzip) base.through(fs2.compress.gunzip(2 * 8196)) else base
+            }
+
+            val writePath = (tmp: File, partNumber: Long, outPrefix: File) => {
+              val outFile = outPrefix / s"part-$partNumber.json"
+              contextShift.evalOn(ec)(IO.delay(tmp.copyTo(outFile))).void
+            }
+
+            new XmlExtractor[File](readPath, writePath)
+          }
+
+          for {
+            _ <- IO.delay(out.createDirectories())
+            _ <- extractor.extract(in, out, tag, count)
+          } yield {
+            ExitCode.Success
```

Comment:
Do we want in indication of done or an indication of number of files written at any given time?